### DISCUSSION TITLE
Can we use javascript classes at server side

### DISCUSSION THREAD

**Discussion by zubair999** [vercel/next.js#73304] [Help]
### Summary

I want to implement OOP in nextjs like I want to implement backend logic in javascript classes

### Additional information

```js
I want to implement OOP in nextjs 
like I want to implement backend logic in javascript classes

like

export class OrderServices implement OrderServiceInterface {

}

like for payment gatways

// some imports

export class GatewayPayService {
  gateway;

  constructor() {
    const keyId = process.env._KEY_ID;
    const secret = process.env._SECRET;

    if (!keyId || !secret) {
      throw new Error("keys are missing from environment variables");
    }

    this.gateway = new SomeGatewayClass({
      key_id: keyId,
      key_secret: secret,
    });
  }

  public async createSomePaymentGatewayOrder(body: Order) {
    const { total, currency, receipt } = body;

    try {
    } catch (err) {}
  }

  public async verifySomePaymentGatewayOrder(body: Order) {
    try {
    } catch (error) {}
  }
}


like this

is safe and ok because i want to use this only in server side and when we use "use server",
it say's only async function, this should also support

However I am using classes and it is working but I want to know is this ok for nextjs for security reason

env: {
    KEY_ID: process.env.KEY_ID,
    SECRET: process.env.SECRET,
  },
  
  i have set this in next config file and using
```

**Comment by denis23x** [comment]
Hi @zubair999

Yes, I think so.., there would be security problems if you used that keys in the client application, since attackers can parse your bundle and steal the keys.

**Reply by zubair999** [reply]
`"use server";

import { Order } from "@/types";
import { OrderServices } from "@/actions/order-services";

const orderServices = new OrderServices();

// server action
export const addOrder = async (
  data: Order
): Promise<{ status: string; message: string }> => {
  try {
    const response = await orderServices.addOrder(data);
    return { status: "success", message: "Order added successfully" };
  } catch (error) {
    console.error("Error in addOrder:", error);
    throw new Error("Something went wrong while processing the order.");
  }
};


// order services class in different file
export class OrderServices implements OrderServiceInterface {
}`


right now this is working but I am not sure is this safe or not because we can't use use server here.
same i want to implement for other like gateways.

**Comment by icyJoseph** [comment]
- https://nextjs.org/blog/security-nextjs-server-components-actions#closures

> The snapshot of the closure must be sent to the client and back when the server is invoked.
> 
> In Next.js 14, the closed over variables are encrypted with the action ID before sent to the client. By default a private key is generated automatically during the build of a Next.js project. Each rebuild generates a new private key which means that each Server Action can only be invoked for a specific build. You might want to use [Skew Protection](https://vercel.com/blog/version-skew-protection) to ensure that you always invoke the correction version during redeploys.
> 
> If you need a key that rotates more frequently or is persistent across multiple builds, you can configure it manually using NEXT_SERVER_ACTIONS_ENCRYPTION_KEY environment variable.
> 
> By encrypting all closed over variables, you don't accidentally expose any secrets in them. By signing it, it makes it harder for an attacker to mess with the input to the action.

You should be fine.

I'd recommend installing `server-only` as well, and do:

```ts
import 'server-only';

// your OOP stuff
export class OrderServices implement OrderServiceInterface {

}
```

This is an additional measure to prevent, accidental imports into client code.
